File: cpp/src/paths.cpp
#include "rohrpost/paths.hpp"

#include "rohrpost/errors.hpp"
#include "rohrpost/io.hpp"

#include <algorithm>
#include <array>
#include <cstddef>
#include <filesystem>
#include <optional>
#include <string>
#include <string_view>
#include <system_error>
#include <utility>
#include <vector>

namespace rp::paths {
namespace {

namespace fs = std::filesystem;

fs::path resolve(const fs::path& p) {
    std::error_code ec;
    fs::path out = fs::weakly_canonical(p, ec);
    return ec ? fs::absolute(p) : out;
}

/// `[here, *here.parents]`.
std::vector<fs::path> ancestry(const fs::path& here) {
    std::vector<fs::path> out{here};
    fs::path current = here;
    for (;;) {
        const fs::path parent = current.parent_path();
        if (parent.empty() || parent == current) break;
        out.push_back(parent);
        current = parent;
    }
    return out;
}

/// Python's `line not in existing` after universal-newline decoding.
std::string normalise_newlines(std::string_view text) {
    std::string out;
    out.reserve(text.size());
    for (std::size_t i = 0; i < text.size(); ++i) {
        if (text[i] == '\r') {
            out.push_back('\n');
            if (i + 1 < text.size() && text[i + 1] == '\n') ++i;
        } else {
            out.push_back(text[i]);
        }
    }
    return out;
}

/// Append any of `lines` not already present to `path`. Returns whether changed.
template <std::size_t N>
bool append_unique_lines(const fs::path& path, const std::array<std::string_view, N>& lines) {
    std::string existing;
    std::error_code ec;
    if (fs::is_regular_file(path, ec)) {
        auto read = io::read_file(path);
        if (!read) throw StoreError("cannot read " + io::path_str(path) + ": " + read.error());
        existing = normalise_newlines(*read);
    }
    std::vector<std::string_view> fresh;
    for (const auto line : lines) {
        if (existing.find(line) == std::string::npos) fresh.push_back(line);
    }
    if (fresh.empty()) return false;
    std::string payload;
    if (!existing.empty() && !existing.ends_with('\n')) payload.push_back('\n');
    for (std::size_t i = 0; i < fresh.size(); ++i) {
        if (i > 0) payload.push_back('\n');
        payload.append(fresh[i]);
    }
    payload.push_back('\n');
    io::append_file(path, payload);
    return true;
}

}  // namespace

fs::path resolved_cwd() {
    return resolve(fs::current_path());
}

std::optional<fs::path> find_git_root(std::optional<fs::path> start) {
    const fs::path here = start ? resolve(*start) : resolved_cwd();
    for (const auto& candidate : ancestry(here)) {
        std::error_code ec;
        if (fs::exists(candidate / ".git", ec)) return candidate;
    }
    return std::nullopt;
}

std::optional<fs::path> find_rohrpost_dir(std::optional<fs::path> start) {
    const fs::path here = start ? resolve(*start) : resolved_cwd();
    for (const auto& candidate : ancestry(here)) {
        std::error_code ec;
        if (fs::is_directory(candidate / kRohrpostDirName, ec)) return candidate / kRohrpostDirName;
    }
    return std::nullopt;
}

fs::path require_rohrpost_dir(std::optional<fs::path> start) {
    auto found = find_rohrpost_dir(std::move(start));
    if (!found) throw StoreError("not a rohrpost repository (no .rohrpost/ found). Run `rp init` first.");
    return *found;
}

fs::path config_path(const fs::path& d) { return d / kConfigFilename; }
fs::path log_path(const fs::path& d) { return d / kLogFilename; }
fs::path snapshot_path(const fs::path& d) { return d / kSnapshotFilename; }
fs::path archive_dir(const fs::path& d) { return d / kArchiveDirName; }
fs::path shadow_dir(const fs::path& d) { return d / kShadowDirName; }
fs::path templates_dir(const fs::path& d) { return d / kTemplatesDirName; }
fs::path bodies_dir(const fs::path& d) { return d / kBodiesDirName; }
fs::path lock_path(const fs::path& d) { return d / kLockFilename; }

std::vector<fs::path> archive_files(const fs::path& rohrpost_dir) {
    const fs::path adir = archive_dir(rohrpost_dir);
    std::error_code ec;
    if (!fs::is_directory(adir, ec)) return {};
    std::vector<fs::path> out;
    for (const auto& entry : fs::directory_iterator(adir, ec)) {
        if (entry.path().extension() == ".jsonl") out.push_back(entry.path());
    }
    std::sort(out.begin(), out.end());
    return out;
}

void ensure_layout(const fs::path& rohrpost_dir) {
    std::error_code ec;
    for (const auto& d : {rohrpost_dir, archive_dir(rohrpost_dir), shadow_dir(rohrpost_dir), templates_dir(rohrpost_dir)}) {
        fs::create_directories(d, ec);
        if (ec) throw StoreError("cannot create " + io::path_str(d) + ": " + ec.message());
    }
    const fs::path log = log_path(rohrpost_dir);
    if (!fs::exists(log, ec)) io::write_file(log, "");
}

bool write_gitattributes(const fs::path& repo_root) {
    return append_unique_lines(repo_root / ".gitattributes", kGitattributesRules);
}

bool write_gitignore(const fs::path& repo_root) {
    return append_unique_lines(repo_root / ".gitignore", kGitignoreRules);
}

}  // namespace rp::paths

File: src/paths.rs
//! The `.rohrpost/` layout (spec §4) and repository discovery.
//!
//! Discovery walks up from the current directory looking for `.rohrpost/`, the
//! way git finds `.git/`, so `rp` works from anywhere inside a repo. This is
//! the single place that builds these paths.

use std::path::{Path, PathBuf};

use crate::error::{Error, Result, io_error};

pub const ROHRPOST_DIR_NAME: &str = ".rohrpost";
pub const CONFIG_FILENAME: &str = "config.toml";
pub const LOG_FILENAME: &str = "log.jsonl";
pub const ARCHIVE_DIR_NAME: &str = "archive";
pub const TEMPLATES_DIR_NAME: &str = "templates";
pub const LOCK_FILENAME: &str = ".lock";

/// The committed `.gitattributes` rules (spec §4). `merge=union` keeps both
/// sides' appended lines; `text eol=lf` keeps the JSONL byte-identical across
/// platforms regardless of `core.autocrlf`.
pub const GITATTRIBUTES_RULES: &[&str] = &[
    ".rohrpost/log.jsonl          merge=union text eol=lf",
    ".rohrpost/archive/*.jsonl    merge=union text eol=lf",
];

pub fn config_path(dir: &Path) -> PathBuf {
    dir.join(CONFIG_FILENAME)
}
pub fn log_path(dir: &Path) -> PathBuf {
    dir.join(LOG_FILENAME)
}
pub fn archive_dir(dir: &Path) -> PathBuf {
    dir.join(ARCHIVE_DIR_NAME)
}
pub fn templates_dir(dir: &Path) -> PathBuf {
    dir.join(TEMPLATES_DIR_NAME)
}
pub fn lock_path(dir: &Path) -> PathBuf {
    dir.join(LOCK_FILENAME)
}

/// The repository root that owns a `.rohrpost/` dir (its parent).
pub fn repo_root(rohrpost_dir: &Path) -> &Path {
    rohrpost_dir.parent().unwrap_or(rohrpost_dir)
}

fn current_dir() -> Result<PathBuf> {
    std::env::current_dir()
        .map_err(|e| Error::Store(format!("cannot determine the current directory: {e}")))
}

/// Nearest ancestor of `start` (default cwd) containing `.git`, if any.
pub fn find_git_root(start: Option<&Path>) -> Result<Option<PathBuf>> {
    let here = match start {
        Some(p) => p.to_path_buf(),
        None => current_dir()?,
    };
    Ok(here
        .ancestors()
        .find(|dir| dir.join(".git").exists())
        .map(Path::to_path_buf))
}

/// Nearest ancestor of `start` (default cwd) containing `.rohrpost/`, if any.
pub fn find_rohrpost_dir(start: Option<&Path>) -> Result<Option<PathBuf>> {
    let here = match start {
        Some(p) => p.to_path_buf(),
        None => current_dir()?,
    };
    Ok(here
        .ancestors()
        .map(|dir| dir.join(ROHRPOST_DIR_NAME))
        .find(|candidate| candidate.is_dir()))
}

/// The `.rohrpost/` dir for the cwd, or a "run `rp init`" error.
pub fn require_rohrpost_dir() -> Result<PathBuf> {
    find_rohrpost_dir(None)?.ok_or_else(|| {
        Error::Store("not a rohrpost repository (no .rohrpost/ found). Run `rp init` first.".into())
    })
}

/// Sorted `archive/*.jsonl` files (oldest first by name). Empty if none.
pub fn archive_files(dir: &Path) -> Result<Vec<PathBuf>> {
    let adir = archive_dir(dir);
    if !adir.is_dir() {
        return Ok(Vec::new());
    }
    let mut files: Vec<PathBuf> = std::fs::read_dir(&adir)
        .map_err(|e| io_error("cannot list", &adir, &e))?
        .filter_map(|entry| entry.ok().map(|e| e.path()))
        .filter(|p| p.is_file() && p.extension().is_some_and(|ext| ext == "jsonl"))
        .collect();
    files.sort();
    Ok(files)
}

/// Create the directory scaffold and an empty log if missing (idempotent).
pub fn ensure_layout(dir: &Path) -> Result<()> {
    for d in [dir.to_path_buf(), archive_dir(dir), templates_dir(dir)] {
        std::fs::create_dir_all(&d).map_err(|e| io_error("cannot create", &d, &e))?;
    }
    let log = log_path(dir);
    std::fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(&log)
        .map_err(|e| io_error("cannot create", &log, &e))?;
    Ok(())
}

/// Append any of `lines` not already present to `path` (LF, UTF-8 bytes).
/// Returns whether the file changed. The file is user-owned and may predate
/// `rp`, so it is searched and extended as raw bytes, never re-encoded.
fn append_unique_lines(path: &Path, lines: &[&str]) -> Result<bool> {
    let existing = match std::fs::read(path) {
        Ok(bytes) => bytes,
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => Vec::new(),
        Err(e) => return Err(io_error("cannot read", path, &e)),
    };
    let contains = |needle: &str| {
        existing
            .windows(needle.len())
            .any(|w| w == needle.as_bytes())
    };
    let missing: Vec<&str> = lines
        .iter()
        .copied()
        .filter(|line| !contains(line))
        .collect();
    if missing.is_empty() {
        return Ok(false);
    }
    let mut out = Vec::new();
    if !existing.is_empty() && !existing.ends_with(b"\n") {
        out.push(b'\n');
    }
    out.extend_from_slice(missing.join("\n").as_bytes());
    out.push(b'\n');
    use std::io::Write as _;
    let mut file = std::fs::OpenOptions::new()
        .append(true)
        .create(true)
        .open(path)
        .map_err(|e| io_error("cannot open", path, &e))?;
    file.write_all(&out)
        .map_err(|e| io_error("cannot write", path, &e))?;
    Ok(true)
}

/// Ensure the committed `.gitattributes` carries the merge and eol rules.
/// Idempotent; later lines win in gitattributes, so appending upgrades a file
/// that carries older rules.
pub fn write_gitattributes(repo_root: &Path) -> Result<bool> {
    append_unique_lines(&repo_root.join(".gitattributes"), GITATTRIBUTES_RULES)
}
